grpc/src/main/java/io/stargate/grpc/service/MessageHandler.java

```diff
@@ -0,0 +1,305 @@
+ * limitations under the License.
+ */
+package io.stargate.grpc.service;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.StatusException;
+import io.grpc.StatusRuntimeException;
+import io.grpc.protobuf.ProtoUtils;
+import io.grpc.stub.StreamObserver;
+import io.stargate.db.ImmutableParameters;
+import io.stargate.db.Parameters;
+import io.stargate.db.Persistence;
+import io.stargate.db.Persistence.Connection;
+import io.stargate.db.Result.Prepared;
+import io.stargate.grpc.service.Service.PrepareInfo;
+import io.stargate.proto.QueryOuterClass;
+import io.stargate.proto.QueryOuterClass.Response;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import org.apache.cassandra.stargate.exceptions.AlreadyExistsException;
+import org.apache.cassandra.stargate.exceptions.CasWriteUnknownResultException;
+import org.apache.cassandra.stargate.exceptions.FunctionExecutionException;
+import org.apache.cassandra.stargate.exceptions.PersistenceException;
+import org.apache.cassandra.stargate.exceptions.ReadFailureException;
+import org.apache.cassandra.stargate.exceptions.ReadTimeoutException;
+import org.apache.cassandra.stargate.exceptions.UnavailableException;
+import org.apache.cassandra.stargate.exceptions.WriteFailureException;
+import org.apache.cassandra.stargate.exceptions.WriteTimeoutException;
+
+abstract class MessageHandler {
+
+  private static final Metadata.Key<QueryOuterClass.Unavailable> UNAVAILABLE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.Unavailable.getDefaultInstance());
+  private static final Metadata.Key<QueryOuterClass.WriteTimeout> WRITE_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteTimeout.getDefaultInstance());
+  private static final Metadata.Key<QueryOuterClass.ReadTimeout> READ_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadTimeout.getDefaultInstance());
+  private static final Metadata.Key<QueryOuterClass.ReadFailure> READ_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadFailure.getDefaultInstance());
+  private static final Metadata.Key<QueryOuterClass.FunctionFailure> FUNCTION_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.FunctionFailure.getDefaultInstance());
+  private static final Metadata.Key<QueryOuterClass.WriteFailure> WRITE_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteFailure.getDefaultInstance());
+  private static final Metadata.Key<QueryOuterClass.AlreadyExists> ALREADY_EXISTS_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.AlreadyExists.getDefaultInstance());
+  private static final Metadata.Key<QueryOuterClass.CasWriteUnknown> CAS_WRITE_UNKNOWN_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.CasWriteUnknown.getDefaultInstance());
+
+  protected final Connection connection;
+  private final Cache<PrepareInfo, CompletionStage<Prepared>> preparedCache;
```

Comment:
Storing futures in the cache allows us to prepare exactly once if multiple threads are racing on the same statement (see below).

Reply:
Nice fix.